### PR TITLE
Document the centreline data

### LINKED ISSUE #801: We need better documentation for the partitioned tables in `gis_core`
How should people be using them?
Can we have a simple way of accessing the latest data? A Mat View maybe?

### PR DESCRIPTION
**Pull request by Nate-Wessel**
## What this pull request accomplishes:

- documents the centreline data

## Issue(s) this solves:

closes #801

## What, in particular, needs to reviewed:

- facts, clarity

## What needs to be done by a sysadmin after this PR is merged

NA

### CODE REVIEW

**Comment by Nate-Wessel** [comment]
@chmnata This is not meant to be complete; I've only been able to find out so much about this dataset, and currently there is no documentation that I'm aware of. This is just intended to be a start that we can add to later as time permits. Thanks!

File: gis/centreline/readme.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+The centreline data is structured as an undirected graph with edges and nodes. Both edges and nodes have a `centreline_id` identifier. A given `centreline_id` will refer to either an edge or a node. All edges have _from_ and _to_ nodes, though this should not be taken to indicate that edges are directed. A two-way street will be represented by a single edge and the _from_ and _to_ identifiers may be arbitrarily swapped.
+
+# Where It's Stored
+
+Centreline data are stored in the `gis_core` schema in the `bigdata` database. Nodes are stored separately from edges. There may be older copy of the various centreline tables in the `gis` schema but they are not updated and their use is deprecated.
+
+## Edges
+
+Edges are stored in the partitioned table `gis_core.centreline`. This is not the complete centreline dataset - it has had some pre-filtering done to it to remove things like administrative boundaries.
```
Comment (chmnata) [inline gis/centreline/readme.md:17]: Currently we are filtering for 'Collector','Collector Ramp','Expressway','Expressway Ramp','Local','Major Arterial','Major Arterial Ramp','Minor Arterial','Minor Arterial Ramp','Pending'. But we will be pulling `Other` soon!

File: gis/centreline/readme.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+However not all transport related edges are included. Things like alleyways and footpaths have been filtered out.
+
+## Nodes
+
+Nodes are stored in either of two tables, each of which is copied from a separate GCC layer which in turn are maintained by different groups. The differences between these tables is not completely understood.
```
Comment (chmnata) [inline gis/centreline/readme.md:23]: FYI @raha1986 looked into `intersection` before https://github.com/CityofToronto/bdit_data-sources/issues/466. We could also include some comments mentioned on centreline intersections e.g. elevation10 can be used to filter major and minor arterials, and limitations like how centreline id for intersections are not unique

**Review by chmnata — CHANGES_REQUESTED** [review]
Thanks for starting a readme for centreline! Some comments below, and I will have to update it after merging to include some new changes on centreline_latest.

File: gis/centreline/readme.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# City Centreline Data
+
+The centreline data is one of the City's core GIS products. It delineates street centrelines as well as some other entity types like waterways or political boundaries. For the most part, our unit will need to filter out certain classes of these other entity types which are not directy related to transportation.
+
+The centreline data are used by many other groups in the City and it's often important to be able to relate various data products or infrastucture back to the centreline entities they describe or are located on.
+
+## How It's Structured
+
+The centreline data is structured as an undirected graph with edges and nodes. Both edges and nodes have a `centreline_id` identifier. A given `centreline_id` will refer to either an edge or a node. All edges have _from_ and _to_ nodes, though this should not be taken to indicate that edges are directed. A two-way street will be represented by a single edge and the _from_ and _to_ identifiers may be arbitrarily swapped.
```
[the remark this answers is not in the document]
Comment (Nate-Wessel) [reply gis/centreline/readme.md:9]: Assuming this is for cars only? Or would it account for bicycle contra-flow lanes as well?

[the remark this answers is not in the document]
Comment (Nate-Wessel) [reply gis/centreline/readme.md]: I've added a note about directed edges and `oneway_dir_code`

File: gis/centreline/readme.md (hunk 1 above)
Comment (Nate-Wessel) [reply gis/centreline/readme.md:17]: Added!

File: gis/centreline/readme.md (hunk 2 above)
Comment (Nate-Wessel) [reply gis/centreline/readme.md:41]: I'm not sure I'll be able to give a good summary of the discussion there. Perhaps we could let Raha add to this documentation in a later PR?

**Comment by chmnata** [comment]
Thanks @Nate-Wessel for the speedy fixes, @radumas wanna take a quick look at it? FYI I am gonna add the routing stuff in after this merge.

**Review by radumas — CHANGES_REQUESTED** [review]
I realize there's a `centreline` folder but I kiiiind of find it not that useful (anymore). Feel like this should instead be front and centre in the `gis/README.md`, though maybe it's a bit long. IDK.

There should be a reference somewhere that the code for the pipelines for these datasets are in `gccview/` folder.

`centreline_intersection_point`  I would say is the preferred intersection layer, it is **almost** unique on `intersection_id`

image

`intersection` appears to be 1 row to describe **every** relationship between edges at a node.

I think it would be helpful to add that `intersection_id` is the ID in the intersection layer(s) that matches `centreline_id` from MOVE

File: gis/centreline/readme.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+### Intersections (nodes)
+
+Intersections are stored in either of two tables, each of which is copied from a separate GCC layer which in turn are maintained by different groups. Each intersections represents the intersecting point of two or more centreline segments.
+
+* `gis_core.intersection` (pulled from [here](https://insideto-gis.toronto.ca/arcgis/rest/services/cot_geospatial12/FeatureServer/42))
+    - contains additional elevation information such as elevation level, elevation unit, height restriction, etc
+    - does not include cul-de-sacs, overpass/underpass
+* `gis_core.centreline_intersection_point` (pulled from [here](https://insideto-gis.toronto.ca/arcgis/rest/services/cot_geospatial/FeatureServer/19))
+    - contains additional boundary information such as ward, and municpality
+    - include trails and ferry routes
```
Comment (gabrielwol) [inline gis/centreline/readme.md:58]: Can we add these as comments on the sql tables to guide people when to use them?

**Review by gabrielwol — CHANGES_REQUESTED** [review]
I appreciate this attempt to document the centreline. 
As a centreline user recently what I feel is still lacking is some obvious pointers to use `gis_core` over `gis`, and `centreline_latest` over `centreline`.
- As part of this PR could we add comments to these two schemas and most of the tables under `gis_core`?
- Add a link and overview of this readme under the main gis/readme.md.

File: gis/centreline/readme.md (hunk 4 above)
Comment (chmnata) [reply gis/centreline/readme.md:58]: Good Call. Added comment for `gis.centreline`, as well as the two above intersection tables.

Comment (chmnata) [reply gis/centreline/readme.md:58]: I tried to not mention `gis` schema at all and only mention `gis_core` in the How It's Structured block. What other pointers should we include to make the distinction?

**Review by gabrielwol — APPROVED** [review]
Great improvements to the documentation. Also thanks for adding comments on those tables in bigdata.

**Comment by gabrielwol** [comment]
@radumas I addressed your comments from above. Can you review updates?

**Review by chmnata — APPROVED** [review]
Thanks for adding all the other folders description and more info on intersection! @gabrielwol

**Review by radumas — APPROVED** [review]
Great job team

**Pull request merged**